### ISSUE TITLE
fails to build on warp-ctc

### ISSUE THREAD

**Issue by Nickerchen** [espnet/espnet#703]
Hello everyone, 
the make process fails during the installation of warp-ctc and I am not sure why, all the dependencies should be there.

```
(kaldi) jamfri@pc318linux:~/Downloads/espnet/tools$ make KALDI=/home/jamfri/Downloads/kaldi
rm -rf warp-ctc
git clone https://github.com/jnishi/warp-ctc.git
Cloning into 'warp-ctc'...
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 811 (delta 0), reused 2 (delta 0), pack-reused 807
Receiving objects: 100% (811/811), 384.88 KiB | 977.00 KiB/s, done.
Resolving deltas: 100% (410/410), done.
# Note(kamo): Requires gcc>=4.9 to build extensions with pytorch>=1.0
if . venv/bin/activate && python -c 'import torch as t;assert t.__version__[0] == "1"' &> /dev/null; then \
        . venv/bin/activate && python -c "from distutils.version import LooseVersion as V;assert V('5.5.0') >= V('4.9'), 'Requires gcc>=4.9'"; \
fi
if . venv/bin/activate && python -c 'import torch as t;assert t.__version__[0] == "1"' &> /dev/null; then \
        cd warp-ctc; git checkout -b pytorch-1.0.0 remotes/origin/pytorch-1.0.0; \
    fi
Branch 'pytorch-1.0.0' set up to track remote branch 'pytorch-1.0.0' from 'origin'.
Switched to a new branch 'pytorch-1.0.0'
. venv/bin/activate; cd warp-ctc && mkdir build && cd build && cmake .. && make -j4 ; true
-- The C compiler identification is GNU 5.5.0
-- The CXX compiler identification is GNU 7.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found CUDA: /usr/local/cuda (found suitable version "10.1", minimum required is "6.5") 
-- cuda found TRUE
-- Building shared library with GPU support
-- Configuring done
-- Generating done
-- Build files have been written to: /home/jamfri/Downloads/espnet/tools/warp-ctc/build
make[1]: Entering directory '/home/jamfri/Downloads/espnet/tools/warp-ctc/build'
make[2]: Entering directory '/home/jamfri/Downloads/espnet/tools/warp-ctc/build'
make[3]: Entering directory '/home/jamfri/Downloads/espnet/tools/warp-ctc/build'
[ 11%] Building NVCC (Device) object CMakeFiles/warpctc.dir/src/warpctc_generated_ctc_entrypoint.cu.o
[ 22%] Building NVCC (Device) object CMakeFiles/warpctc.dir/src/warpctc_generated_reduce.cu.o
/home/jamfri/Downloads/espnet/tools/warp-ctc/src/reduce.cu(44): error: identifier "__shfl_down" is undefined
          detected during:
            instantiation of "T CTAReduce<NT, T, Rop>::reduce(int, T, CTAReduce<NT, T, Rop>::Storage &, int, Rop) [with NT=128, T=float, Rop=ctc_helper::add<float, float>]" 
(76): here
            instantiation of "void reduce_rows<NT,Iop,Rop,T>(Iop, Rop, const T *, T *, int, int) [with NT=128, Iop=ctc_helper::negate<float, float>, Rop=ctc_helper::add<float, float>, T=float]" 
(124): here
            instantiation of "void ReduceHelper::impl(Iof, Rof, const T *, T *, int, int, __nv_bool, cudaStream_t) [with T=float, Iof=ctc_helper::negate<float, float>, Rof=ctc_helper::add<float, float>]" 
(139): here
            instantiation of "ctcStatus_t reduce(Iof, Rof, const T *, T *, int, int, __nv_bool, cudaStream_t) [with T=float, Iof=ctc_helper::negate<float, float>, Rof=ctc_helper::add<float, float>]" 
(149): here

/home/jamfri/Downloads/espnet/tools/warp-ctc/src/reduce.cu(44): error: identifier "__shfl_down" is undefined
          detected during:
            instantiation of "T CTAReduce<NT, T, Rop>::reduce(int, T, CTAReduce<NT, T, Rop>::Storage &, int, Rop) [with NT=128, T=float, Rop=ctc_helper::maximum<float, float>]" 
(76): here
            instantiation of "void reduce_rows<NT,Iop,Rop,T>(Iop, Rop, const T *, T *, int, int) [with NT=128, Iop=ctc_helper::identity<float, float>, Rop=ctc_helper::maximum<float, float>, T=float]" 
(124): here
            instantiation of "void ReduceHelper::impl(Iof, Rof, const T *, T *, int, int, __nv_bool, cudaStream_t) [with T=float, Iof=ctc_helper::identity<float, float>, Rof=ctc_helper::maximum<float, float>]" 
(139): here
            instantiation of "ctcStatus_t reduce(Iof, Rof, const T *, T *, int, int, __nv_bool, cudaStream_t) [with T=float, Iof=ctc_helper::identity<float, float>, Rof=ctc_helper::maximum<float, float>]" 
(157): here

2 errors detected in the compilation of "/tmp/tmpxft_00001937_00000000-13_reduce.compute_70.cpp1.ii".
CMake Error at warpctc_generated_reduce.cu.o.cmake:279 (message):
  Error generating file
  /home/jamfri/Downloads/espnet/tools/warp-ctc/build/CMakeFiles/warpctc.dir/src/./warpctc_generated_reduce.cu.o


CMakeFiles/warpctc.dir/build.make:70: recipe for target 'CMakeFiles/warpctc.dir/src/warpctc_generated_reduce.cu.o' failed
make[3]: *** [CMakeFiles/warpctc.dir/src/warpctc_generated_reduce.cu.o] Error 1
make[3]: *** Waiting for unfinished jobs....
/home/jamfri/Downloads/espnet/tools/warp-ctc/include/contrib/moderngpu/include/device/intrinsics.cuh(115): error: identifier "__shfl_up" is undefined

/home/jamfri/Downloads/espnet/tools/warp-ctc/include/contrib/moderngpu/include/device/intrinsics.cuh(125): error: identifier "__shfl_up" is undefined

2 errors detected in the compilation of "/tmp/tmpxft_00001949_00000000-13_ctc_entrypoint.compute_70.cpp1.ii".
CMake Error at warpctc_generated_ctc_entrypoint.cu.o.cmake:279 (message):
  Error generating file
  /home/jamfri/Downloads/espnet/tools/warp-ctc/build/CMakeFiles/warpctc.dir/src/./warpctc_generated_ctc_entrypoint.cu.o


CMakeFiles/warpctc.dir/build.make:63: recipe for target 'CMakeFiles/warpctc.dir/src/warpctc_generated_ctc_entrypoint.cu.o' failed
make[3]: *** [CMakeFiles/warpctc.dir/src/warpctc_generated_ctc_entrypoint.cu.o] Error 1
make[3]: Leaving directory '/home/jamfri/Downloads/espnet/tools/warp-ctc/build'
CMakeFiles/Makefile2:104: recipe for target 'CMakeFiles/warpctc.dir/all' failed
make[2]: *** [CMakeFiles/warpctc.dir/all] Error 2
make[2]: Leaving directory '/home/jamfri/Downloads/espnet/tools/warp-ctc/build'
Makefile:129: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/home/jamfri/Downloads/espnet/tools/warp-ctc/build'
. venv/bin/activate; pip install cffi
Requirement already satisfied: cffi in ./venv/lib/python3.7/site-packages (1.12.2)
Requirement already satisfied: pycparser in ./venv/lib/python3.7/site-packages (from cffi) (2.19)
. venv/bin/activate; cd warp-ctc/pytorch_binding && python setup.py install # maybe need to: apt-get install python-dev
Could not find libwarpctc.so in ../build.
Build warp-ctc and set WARP_CTC_PATH to the location of libwarpctc.so (default is '../build')
Makefile:90: recipe for target 'warp-ctc.done' failed
make: *** [warp-ctc.done] Error 1

```

**Comment by jnishi** [comment]
It seems the same error as https://github.com/SeanNaren/warp-ctc/issues/116.
It is suggested that some lines of `CMakelists.txt`  to be deleted. I applied them to ` pytiorch-1.0.0-gcc-6` branch  (https://github.com/jnishi/warp-ctc/tree/pytorch-1.0.0-gcc-6). Please check it by fixing https://github.com/espnet/espnet/blob/master/tools/Makefile#L97 this line.

**Comment by Nickerchen** [comment]
thank you for the fast reply!
changing the line
`cd warp-ctc; git checkout -b pytorch-1.0.0 remotes/origin/pytorch-1.0.0; \`
to
`cd warp-ctc; git checkout -b pytorch-1.0.0-gcc-6 remotes/origin/pytorch-1.0.0-gcc-6; \`
fixed it

**Comment by jnishi** [comment]
@Nickerchen Hi, did you check the result of decoding with `pytorch-1.0.0-gcc-6` branch? In my environment,  decoding output is bad in CSJ and Librispeech datasets with the branch, though loss becomes smaller in training.
An example of decode output is like this.

```
id: (8842_302196-8842-302196-0003)
Scores: (#C #S #D #I) 27 1 0 92
REF:  the only true motive for putting poetry into a fresh language must be to endow a fresh nation as far as possible with one more POSSESSION of *** beauty ** *** *** ** *********** ***** ** ******** * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
HYP:  the only true motive for putting poetry into a fresh language must be to endow a fresh nation as far as possible with one more VIEW       of THE beauty OF THE ART OF BEAUCHAMP'S ADIEU TO BEAUTIFY V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V V
Eval:                                                                                                                                S             I          I  I   I   I  I           I     I  I        I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I
```
I checked to run correctly with built-in CTC in pytorch.
Did it run correctly in your environment?
